File: Stack.java
import java.util.LinkedList;

public class Stack <T> {

    private static final int MAX_SIZE = 10000;

    private int maxSize;
    private T top;
    private int size;
    private LinkedList<T> list;

    public Stack(int maxSize) {
        this.maxSize = maxSize;
        this.size = 0;
        this.top = null;
        this.list = new LinkedList<T>();
    }

    public void push(T data) {
        if (!isFull()) {
            this.list.addFirst(data);
            this.top = data;
            size++;
        }
    }

    public T pop() {
        if (isEmpty()) {
            return null;
        } else {
            T data = this.list.removeFirst();
            this.top = this.list.getFirst();
            size--;
            return data;
        } 
    }

    public boolean isEmpty() {
        return size == 0;
    }

    public boolean isFull() {
        return size == maxSize;
    }

    public T peek() {
        return this.list.getFirst();
    }

    public T top() {
        return this.top;
    }

    public int getCapacity() {
        return this.maxSize;
    }
}

File: Stack.js
// Implement a stack in JS

class Stack {
    constructor(maxSize) {
      this._list = [];
      this._maxSize = maxSize !== undefined ? maxSize : 100;
      console.log(this._maxSize);
    }
  
    pop() {
      if (this.isEmpty()) {
        throw new Error('The stack is empty.');
      } else {
        const item = this._list.pop();
        return item;
      }
    }
  
    push(item) {
      if (this.isFull()) {
        throw new Error('The stack is full.');
      } else {
        this._list.push(item);
      }
    }
  
    isTop(item) {
      return this.peek() === item;
    }
  
    size() {
      return this._list.length;
    }
  
    peek() {
      if (this.isEmpty()) {
        throw new Error('The stack is empty.');
      } else {
        return this._list[this._list.length];
      }
    }
  
    isEmpty() {
      return this._list.length <= 0;
    }
  
    isFull() {
      return this._list.length === this._maxSize;
    }
    printStack() {
      console.log(this._list);
    }
  }
  
  let myStack = new Stack(10);
  myStack.printStack();
  myStack.push(3);
  myStack.push(6);
  myStack.push(9);
  myStack.push(12);
  myStack.printStack();
  myStack.pop();
  myStack.pop();
  myStack.printStack();
  myStack.pop();
  myStack.pop();
  console.log(myStack.isEmpty());
  
  
  // Stacks are used to packtrack to previous task or state like in recusive code. Depth First Search
  // also used to store partially completed tasks, when you are exploring two different paths
  // on a graph from a point while doing the shortest path algorithm. 
  
  